--- a/subprojects/mbedtls-mbedtls-3.4.0/meson.build
+++ b/subprojects/mbedtls-mbedtls-3.4.0/meson.build
@@ -55,13 +55,6 @@
     ],
 )
 
-foreach output : mbedtls_libs_output
-    meson.add_install_script('/bin/sh', '-c',
-        ('ln -sf ../../../@0@ ' +
-        '"$MESON_INSTALL_DESTDIR_PREFIX"/@1@/gramine/runtime/glibc/').format(
-            output, get_option('libdir')))
-endforeach
-
 # We rely on the fact that for `mbedtls_gramine` package, we don't need any changes in the default
 # mbedTLS headers
 install_subdir('mbedtls-mbedtls-3.4.0/include/mbedtls', install_dir: get_option('includedir') / 'gramine')

--- a/tools/sgx/ra-tls/meson.build
+++ b/tools/sgx/ra-tls/meson.build
@@ -28,10 +28,6 @@
     install: true,
     install_rpath: get_option('prefix') / get_option('libdir'),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libra_tls_attest.so ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 libra_tls_verify = static_library('ra_tls_verify',
     'ra_tls_verify_common.c',
@@ -44,10 +40,6 @@
     install: true,
     install_rpath: get_option('prefix') / get_option('libdir'),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libra_tls_verify.a ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 libra_tls_verify_dep = declare_dependency(
     link_with: libra_tls_verify,
@@ -66,10 +58,6 @@
     install: true,
     install_rpath: join_paths(get_option('prefix'), get_option('libdir')),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libra_tls_verify_epid.so ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 libsecret_prov_attest = shared_library('secret_prov_attest',
     'secret_prov_attest.c',
@@ -85,10 +73,6 @@
     install: true,
     install_rpath: get_option('prefix') / get_option('libdir'),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libsecret_prov_attest.so ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 libsecret_prov_verify = static_library('secret_prov_verify',
     'secret_prov_verify.c',
@@ -102,10 +86,6 @@
     install: true,
     install_rpath: join_paths(get_option('prefix'), get_option('libdir')),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libsecret_prov_verify.a ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 libsecret_prov_verify_dep = declare_dependency(
     link_with: libsecret_prov_verify,
@@ -125,10 +105,6 @@
     install: true,
     install_rpath: get_option('prefix') / get_option('libdir'),
 )
-meson.add_install_script('/bin/sh', '-c',
-    'ln -sf ../../../libsecret_prov_verify_epid.so ' +
-    '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-        get_option('libdir')))
 
 if dcap
     libra_tls_verify_dcap = shared_library('ra_tls_verify_dcap',
@@ -144,10 +120,6 @@
         install: true,
         install_rpath: get_option('prefix') / get_option('libdir'),
     )
-    meson.add_install_script('/bin/sh', '-c',
-        'ln -sf ../../../libra_tls_verify_dcap.so ' +
-        '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-            get_option('libdir')))
 
     libra_tls_verify_dcap_gramine = shared_library('ra_tls_verify_dcap_gramine',
         'ra_tls_verify_dcap.c',
@@ -163,10 +135,6 @@
         install: true,
         install_rpath: get_option('prefix') / get_option('libdir'),
     )
-    meson.add_install_script('/bin/sh', '-c',
-        'ln -sf ../../../libra_tls_verify_dcap_gramine.so ' +
-        '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-            get_option('libdir')))
 
     libsecret_prov_verify_dcap = shared_library('secret_prov_verify_dcap',
         'ra_tls_verify_dcap.c',
@@ -183,10 +151,6 @@
         install: true,
         install_rpath: get_option('prefix') / get_option('libdir'),
     )
-    meson.add_install_script('/bin/sh', '-c',
-        'ln -sf ../../../libsecret_prov_verify_dcap.so ' +
-        '"$MESON_INSTALL_DESTDIR_PREFIX"/@0@/gramine/runtime/glibc/'.format(
-            get_option('libdir')))
 endif
 
 pkgconfig.generate(
